src/library/scala/collection/ArrayOps.scala

```diff
@@ -560,9 +592,38 @@ final class ArrayOps[A](val xs: Array[A]) extends AnyVal {
     *                 an arbitrary number of times, and must not change the result (e.g., `Nil` for list concatenation,
     *                 0 for addition, or 1 for multiplication).
     *  @param op      a binary operator that must be associative.
-    *  @return        the result of applying the fold operator `op` between all the elements and `z`, or `z` if this array is empty.
-    */
-  @`inline` final def fold[A1 >: A](z: A1)(op: (A1, A1) => A1): A1 = foldLeft(z)(op)
+    *  @return        the result of applying the fold operator `op` between all the elements, or `z` if this array is empty.
+    */
+  def fold[A1 >: A](z: A1)(op: (A1, A1) => A1): A1 = {
+    def f[@specialized(AnyRef, Int, Double, Long, Float, Char, Byte, Short, Boolean, Unit) T](xs: Array[T], op: (Any, Any) => Any, z: Any): Any = {
```

Comment:
Does this make any difference:

~~~scala
def f[@specialized(AnyRef, Int, Double, Long, Float, Char, Byte, Short, Boolean, Unit) T](xs: Array[T], op: (Any, Any) => Any, z: Any): Any = {
      // Start with the last element and run the loop from 0 until length-1. It would be more logical to start with
      // the first element and loop from 1 until length but hotspot performs special optimizations for loops starting
      // at 0 which have a huge impact when the actual folding operation is fast.
      val length = xs.length-1
      if(length >= 0) {
        var v: Any = xs(length)
        var i = 0
        while(i < length) {
          v = op(v, xs(i))
          i += 1
        }
        v
      } else z
    }
~~~

Reply:
It's close to the margin of error but seems to be a little bit faster. I'll update the PR.